### PR TITLE
[Bitcoin] Parse address from script and record balance

### PR DESCRIPTION
**Pull request by jolestar**
## Summary

Parse address from script and record balance

part of #1172 

```bash
curl -H "Content-Type: application/json" -X POST \ 
--data '{"jsonrpc":"2.0","method":"rooch_getStates","params":["/object/0x3::bitcoin_light_client::BitcoinStore", {"decode":true}],"id":1}' \
http://localhost:50051
```
```json
{
   "jsonrpc":"2.0",
   "result":[
      {
         "value":"0xee123ac6dfcb83db04b3c0275451e28054f43c96c5f9984b88717bde0227c4c00000000000000000000000000000000000000000000000000000000000000000010198000000000000007deb8dec5622775bb50a68ab33281b1bfdeddc751052b4ea1907d84295d4b740df0bb28f2b8f1e370f54379ee97876a1005e1d955f1e9e2cd6aceda120388f4009ef5a31278c084d6fe96bcd20eb304154490982721bd62130bb16216d2484c41142e682a0360ed456f132e063526c83f6c6f48d3a32370593864cb74d3155e705b585fab25e52849b5553ffb72803bf0b6768eca781a2101f3f054e3be94d65",
         "value_type":"0x2::object::ObjectEntity<0x3::bitcoin_light_client::BitcoinStore>",
         "decoded_value":{
            "abilities":0,
            "type":"0x2::object::ObjectEntity<0x3::bitcoin_light_client::BitcoinStore>",
            "value":{
               "flag":1,
               "id":"0xee123ac6dfcb83db04b3c0275451e28054f43c96c5f9984b88717bde0227c4c0",
               "owner":"0x0000000000000000000000000000000000000000000000000000000000000000",
               "value":{
                  "abilities":8,
                  "type":"0x3::bitcoin_light_client::BitcoinStore",
                  "value":{
                     "balance":{
                        "abilities":4,
                        "type":"0x2::table::Table<0x3::bitcoin_address::BTCAddress, u64>",
                        "value":{
                           "handle":"0x05b585fab25e52849b5553ffb72803bf0b6768eca781a2101f3f054e3be94d65"
                        }
                     },
                     "blocks":{
                        "abilities":4,
                        "type":"0x2::table::Table<address, 0x3::bitcoin_types::Header>",
                        "value":{
                           "handle":"0x7deb8dec5622775bb50a68ab33281b1bfdeddc751052b4ea1907d84295d4b740"
                        }
                     },
                     "hash_to_height":{
                        "abilities":4,
                        "type":"0x2::table::Table<address, u64>",
                        "value":{
                           "handle":"0x09ef5a31278c084d6fe96bcd20eb304154490982721bd62130bb16216d2484c4"
                        }
                     },
                     "height_to_hash":{
                        "abilities":4,
                        "type":"0x2::table::Table<u64, address>",
                        "value":{
                           "handle":"0xdf0bb28f2b8f1e370f54379ee97876a1005e1d955f1e9e2cd6aceda120388f40"
                        }
                     },
                     "latest_block_height":{
                        "abilities":7,
                        "type":"0x1::option::Option<u64>",
                        "value":{
                           "vec":[
                              "152"
                           ]
                        }
                     },
                     "utxo":{
                        "abilities":4,
                        "type":"0x2::table::Table<0x3::bitcoin_types::OutPoint, 0x3::bitcoin_types::TxOut>",
                        "value":{
                           "handle":"0x1142e682a0360ed456f132e063526c83f6c6f48d3a32370593864cb74d3155e7"
                        }
                     }
                  }
               }
            }
         }
      }
   ],
   "id":1
}
```

```bash
curl -H "Content-Type: application/json" -X POST \
--data '{"jsonrpc":"2.0","method":"rooch_listStates","params":["/table/0x05b585fab25e52849b5553ffb72803bf0b6768eca781a2101f3f054e3be94d65/", null ,null, {"decode":true}],"id":1}' \                  
http://localhost:50051
```
```json
{
   "jsonrpc":"2.0",
   "result":[
      {
         "value":"0xee123ac6dfcb83db04b3c0275451e28054f43c96c5f9984b88717bde0227c4c00000000000000000000000000000000000000000000000000000000000000000010198000000000000007deb8dec5622775bb50a68ab33281b1bfdeddc751052b4ea1907d84295d4b740df0bb28f2b8f1e370f54379ee97876a1005e1d955f1e9e2cd6aceda120388f4009ef5a31278c084d6fe96bcd20eb304154490982721bd62130bb16216d2484c41142e682a0360ed456f132e063526c83f6c6f48d3a32370593864cb74d3155e705b585fab25e52849b5553ffb72803bf0b6768eca781a2101f3f054e3be94d65",
         "value_type":"0x2::object::ObjectEntity<0x3::bitcoin_light_client::BitcoinStore>",
         "decoded_value":{
            "abilities":0,
            "type":"0x2::object::ObjectEntity<0x3::bitcoin_light_client::BitcoinStore>",
            "value":{
               "flag":1,
               "id":"0xee123ac6dfcb83db04b3c0275451e28054f43c96c5f9984b88717bde0227c4c0",
               "owner":"0x0000000000000000000000000000000000000000000000000000000000000000",
               "value":{
                  "abilities":8,
                  "type":"0x3::bitcoin_light_client::BitcoinStore",
                  "value":{
                     "balance":{
                        "abilities":4,
                        "type":"0x2::table::Table<0x3::bitcoin_address::BTCAddress, u64>",
                        "value":{
                           "handle":"0x05b585fab25e52849b5553ffb72803bf0b6768eca781a2101f3f054e3be94d65"
                        }
                     },
                     "blocks":{
                        "abilities":4,
                        "type":"0x2::table::Table<address, 0x3::bitcoin_types::Header>",
                        "value":{
                           "handle":"0x7deb8dec5622775bb50a68ab33281b1bfdeddc751052b4ea1907d84295d4b740"
                        }
                     },
                     "hash_to_height":{
                        "abilities":4,
                        "type":"0x2::table::Table<address, u64>",
                        "value":{
                           "handle":"0x09ef5a31278c084d6fe96bcd20eb304154490982721bd62130bb16216d2484c4"
                        }
                     },
                     "height_to_hash":{
                        "abilities":4,
                        "type":"0x2::table::Table<u64, address>",
                        "value":{
                           "handle":"0xdf0bb28f2b8f1e370f54379ee97876a1005e1d955f1e9e2cd6aceda120388f40"
                        }
                     },
                     "latest_block_height":{
                        "abilities":7,
                        "type":"0x1::option::Option<u64>",
                        "value":{
                           "vec":[
                              "152"
                           ]
                        }
                     },
                     "utxo":{
                        "abilities":4,
                        "type":"0x2::table::Table<0x3::bitcoin_types::OutPoint, 0x3::bitcoin_types::TxOut>",
                        "value":{
                           "handle":"0x1142e682a0360ed456f132e063526c83f6c6f48d3a32370593864cb74d3155e7"
                        }
                     }
                  }
               }
            }
         }
      }
   ],
   "id":1
}
```

### CODE REVIEW

File: crates/rooch-framework/sources/bitcoin_light_client.move [hunk 1]
```diff
@@ -103,11 +103,42 @@ module rooch_framework::bitcoin_light_client{
...
+    fun remove_utxo(btc_store: &mut BitcoinStore, outpoint: OutPoint){
+        if(table::contains(&btc_store.utxo, outpoint)){
+            let txout = table::remove(&mut btc_store.utxo, outpoint);
+            let script_pubkey = bitcoin_types::txout_script_pubkey(&txout);
+            //TODO should we record the address of the utxo?
+            let address_opt = bitcoin_address::from_script(script_pubkey);
+            if(option::is_some(&address_opt)){
+                let address = option::extract(&mut address_opt);
+                let balance = table::borrow_mut_with_default(&mut btc_store.balance, address, 0);
+                if (*balance > bitcoin_types::txout_value(&txout)){
```
Comment (baichuan3) [inline crates/rooch-framework/sources/bitcoin_light_client.move:120]: 1. `>` Should it be `>=`? 
2. In addition, if the balance is not enough, should abort or just emit an event?

Comment (jolestar) [reply crates/rooch-framework/sources/bitcoin_light_client.move]: 1. fixed.
2. Record a comment; I will handle it with the UTXO miss.

**Review by baichuan3 — APPROVED** [review]

File: crates/rooch-framework/sources/bitcoin_network.move [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    ///Currently, Move does not support enum types, so we use constants to represent the network type.    
+    /// Mainnet Bitcoin.
+    const NETWORK_BITCOIN: u8 = 1;
+    public fun network_bitcoin(): u8 {
+        NETWORK_BITCOIN
+    }
+
+    /// Bitcoin's testnet network.
+    const NETWORK_TESTNET: u8 = 2;
```
Comment (feliciss) [inline crates/rooch-framework/sources/bitcoin_network.move:15]: Shall we rename it to NETWORK_BITCOIN_TESTNET for clarity?

File: crates/rooch-framework/sources/bitcoin_network.move [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    public fun network_testnet(): u8 {
+        NETWORK_TESTNET
+    }
+
+    /// Bitcoin's signet network.
+    const NETWORK_SIGNET: u8 = 3;
```
Comment (feliciss) [inline crates/rooch-framework/sources/bitcoin_network.move:21]: There also.

File: crates/rooch-framework/sources/bitcoin_network.move [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        NETWORK_SIGNET
+    }
+
+    /// Bitcoin's regtest network.
+    const NETWORK_REGTEST: u8 = 4;
+    public fun network_regtest(): u8 {
+        NETWORK_REGTEST
+    }
+
+    public fun is_mainnet(network: u8): bool {
```
Comment (feliciss) [inline crates/rooch-framework/sources/bitcoin_network.move:32]: Shall we also add `is_sigtest` or `is_regtest`?

File: crates/rooch-framework/sources/bitcoin_network.move [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        } else if (network == NETWORK_SIGNET) {
+            string::utf8(b"sb")
+        } else if (network == NETWORK_REGTEST) {
+            string::utf8(b"bcrt")
+        } else {
+            abort error::invalid_argument(ErrorUnknownNetwork)
+        }
+    }
+   
+    public fun network_magic(network: u8): vector<u8> {
```
Comment (feliciss) [inline crates/rooch-framework/sources/bitcoin_network.move:78]: What's this?

File: crates/rooch-types/src/address.rs [hunk 6]
```diff
@@ -528,4 +528,21 @@ mod test {
             assert_eq!(address, deserialized);
         }
     }
+
+    #[test]
+    pub fn test_from_script() {
```
Comment (feliciss) [inline crates/rooch-types/src/address.rs:533]: Since it's for signet script test, shall we also add test for regnet or testnet?

**Review by feliciss — COMMENTED** [review]
LGTM.

File: crates/rooch-framework/sources/bitcoin_network.move (hunk 2 above)
Comment (jolestar) [reply crates/rooch-framework/sources/bitcoin_network.move:15]: Read the latest code. it in a bitcoin-move framework.

**Pull request merged**